docs/design/features/typemap.md

```diff
@@ -0,0 +1,223 @@
+    /// Returns the associated type type map generated for the current application.
+    /// </summary>
+    /// <typeparam name="TTypeMapGroup">Type universe</typeparam>
+    /// <param name="map">Requested type map</param>
+    /// <returns>True if the map is returned, otherwise false.</returns>
+    /// <remarks>
+    /// Call sites are treated as an intrinsic by the Trimmer and implemented inline.
+    /// </remarks>
+    [RequiresUnreferencedCode("Interop types may be removed by trimming")]
+    public static IReadOnlyDictionary<Type, Type> GetOrCreateProxyTypeMapping<TTypeMapGroup>();
+}
+```
+
+Given the above types the following would take place.
+
+1. Types involved in unmanaged-to-managed interop operations would be referenced in a
+`TypeMapAttribute` assembly attribute that declared the external type system name, a target
+type, and optionally a "trim-target" to determine if the target
+type should be included in the map. If the `TypeMapAttribute` constructor that doesn't
+take a trim-target is used, the "target type" will be treated as the "trim-target".
+
+2. Types used in a managed-to-unmanaged interop operation would use `TypeMapAssociationAttribute`
+to define a conditional link between the source and proxy type. In other words, if the
+source is kept, so is the proxy type. If Trimmer observes an explicit allocation of the source
+type, the entry will be inserted into the map.
+
+3. During application build, source would be generated and injected into the application
+that defines appropriate `TypeMapAssemblyTargetAttribute` instances. This attribute would help the
+Trimmer know other assemblies to examine for `TypeMapAttribute` and `TypeMapAssociationAttribute`
+instances. These linked assemblies could also be used in the non-Trimmed scenario whereby we
+avoid creating the map at build-time and create a dynamic map at run-time instead.
+
+4. The Trimmer will build two maps based on the above attributes from the application reference
+closure.
+
+    **(a)** Using `TypeMapAttribute` a map from `string` to target `Type`.
+
+    **(b)** Using `TypeMapAssociationAttribute` a map from `Type` to `Type` (source to proxy).
+
+> [!IMPORTANT]
+> Conflicting key/value mappings are not allowed.
+
+> [!NOTE]
+> The underlying format of the produced maps is implementation-defined. Different .NET form factors may use different formats.
+>
+> Additionally, it is not guaranteed that the `TypeMapAttribute`, `TypeMapAssociationAttribute`, and `TypeMapAssemblyTargetAttribute` attributes are present in the final image after a trimming tool has been run.
+
+
+5. Trimming tools will consider calls to `TypeMapping.GetOrCreateExternalTypeMapping<>` and
+`TypeMapping.GetOrCreateProxyTypeMapping<>` as intrinsics (for example, Java via `JavaTypeMapGroup`). As a result, it is not trim-compatible to call either of these methods with non-fully-instantiated generic (such as a type argument or a type that is instantiated over a type argument).
```

Comment:
Is this still true? I see these in the native AOT implementation. Are they expected to be elided in the final binary?

Reply:
The methods are recognized as "intrinsic" method calls in the same way that a number of methods are intrinsics for the linker (they feed into special behavior but are still emitted into the final codegen).